src/persistent_backend_manager.cc

```diff
@@ -0,0 +1,109 @@
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+// OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+#include "persistent_backend_manager.h"
+
+#include <memory>
+#include <mutex>
+#include "backend_config.h"
+#include "filesystem/api.h"
+#include "triton/common/logging.h"
+
+namespace triton { namespace core {
+
+static std::weak_ptr<PersistentBackendManager> persist_backend_manager_;
+static std::mutex mu_;
+
+Status
+PersistentBackendManager::Create(
+    const triton::common::BackendCmdlineConfigMap& config_map,
+    std::shared_ptr<PersistentBackendManager>* manager, std::shared_ptr<TritonBackendManager> backend_manager)
+{
+  std::lock_guard<std::mutex> lock(mu_);
+
+  // If there is already a manager then we just use it...
+  *manager = persist_backend_manager_.lock();
+  if (*manager != nullptr) {
+    return Status::Success;
+  }
+
+  manager->reset(new PersistentBackendManager(backend_manager));
+  persist_backend_manager_ = *manager;
+
+  RETURN_IF_ERROR((*manager)->InitPersistentBackends(config_map));
+  return Status::Success;
+}
+
+Status
+PersistentBackendManager::InitPersistentBackends(
+    const triton::common::BackendCmdlineConfigMap& config_map)
+{
+  for (const auto& be : {"pytorch"}) {
+    RETURN_IF_ERROR(InitPersistentBackend(be, config_map));
+  }
+
+  return Status::Success;
+}
+
+Status
+PersistentBackendManager::InitPersistentBackend(
+    const std::string& backend_name, const triton::common::BackendCmdlineConfigMap& config_map)
+{
+  std::string backends_dir;
+  std::string specialized_backend_name;
+  std::string backend_libname;
+  RETURN_IF_ERROR(
+      BackendConfigurationGlobalBackendsDirectory(config_map, &backends_dir));
+  RETURN_IF_ERROR(BackendConfigurationSpecializeBackendName(
+      config_map, backend_name, &specialized_backend_name));
+  RETURN_IF_ERROR(BackendConfigurationBackendLibraryName(
+      specialized_backend_name, &backend_libname));
+
+  const auto backend_dir = JoinPath({backends_dir, specialized_backend_name});
+  const auto backend_libpath = JoinPath({backend_dir, backend_libname});
+  bool exists = false;
+  RETURN_IF_ERROR(FileExists(backend_libpath, &exists));
+  if (exists) {
+    triton::common::BackendCmdlineConfig empty_backend_cmdline_config;
+    const triton::common::BackendCmdlineConfig* config;
+    const auto& itr = config_map.find(backend_name);
+    if (itr == config_map.end()) {
+      config = &empty_backend_cmdline_config;
+    } else {
+      config = &itr->second;
+    }
+
+    std::shared_ptr<TritonBackend> persist_backend;
+    RETURN_IF_ERROR(backend_manager_->CreateBackend(
```

Comment:
This is the only case I see where that internal backend_manager_ is being used. Why can we not just pass the backend_manager to this method? This is a private method that is only being called by create.

Reply:
Good suggestion, I've the updated the code to include this change.